src/compiler/checker.ts

```diff
@@ -338,7 +338,15 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) ||
+                            !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) ||
+                            lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
Oh so a top level `let` declaration *is* in scope in a parameter, it's just not a legal reference. Whereas local types are not even in scope in the parameters or return type. Is there a reason we chose a different behavior from top level block constructs?
```ts
interface Foo { }
var x: Foo;
function bar(param: Foo = x) { // x reference is an error, but Foo references the outer interface
    interface Foo {
        prop;
    }
    var x: Foo;
}
```